Keen/Dataset/Datasets.cs

```diff
@@ -0,0 +1,301 @@
+
+            var response = JObject.Parse(responseString);
+
+            KeenUtil.CheckApiErrorCode(response);
+
+            if (!responseMessage.IsSuccessStatusCode)
+            {
+                throw new KeenException($"Request failed with status: {responseMessage.StatusCode}");
+            }
+
+            return JsonConvert.DeserializeObject<DatasetDefinition>(responseString, this.GetSerializerSettings());
+        }
+
+        public async Task<DatasetDefinitionCollection> ListDefinitions(int limit = 10, string afterName = null)
+        {
+            if (string.IsNullOrWhiteSpace(_masterKey))
+            {
+                throw new KeenException("An API masterkey is required to get dataset results.");
+            }
+
+            var datasetResultsUrl = $"{_cachedDatasetRelativeUrl}?limit={limit}";
+
+            if (!string.IsNullOrWhiteSpace(afterName))
+            {
+                datasetResultsUrl += $"&after_name={afterName}";
+            }
+
+            var responseMessage = await _keenHttpClient
+                .GetAsync(datasetResultsUrl, _masterKey)
+                .ConfigureAwait( false);
+
+            var responseString = await responseMessage
+                .Content
+                .ReadAsStringAsync()
+                .ConfigureAwait(false);
+
+            var response = JObject.Parse(responseString);
+
+            KeenUtil.CheckApiErrorCode(response);
+
+            if (!responseMessage.IsSuccessStatusCode)
+            {
+                throw new KeenException($"Request failed with status: {responseMessage.StatusCode}");
+            }
+
+            return JsonConvert.DeserializeObject<DatasetDefinitionCollection>(responseString, GetSerializerSettings());
+        }
+
+        public async Task<IEnumerable<DatasetDefinition>> ListAllDefinitions()
+        {
+            var allDefinitions = new List<DatasetDefinition>();
+
+            var firstSet = await this.ListDefinitions(MAX_DATASET_DEFINITION_LIST_LIMIT);
+
+            if (firstSet?.Datasets == null)
+            {
+                throw new KeenException("Failed to fetch definition list");
+            }
+
+            if (!firstSet.Datasets.Any())
+            {
+                return allDefinitions;
+            }
+                
+            if (firstSet.Count <= firstSet.Datasets.Count())
+            {
+                return firstSet.Datasets;
+            }
+            
+            allDefinitions.AddRange(firstSet.Datasets);
+
+            do
+            {
+                var nextSet = await this.ListDefinitions(MAX_DATASET_DEFINITION_LIST_LIMIT, allDefinitions.Last().DatasetName);
+
+                if (nextSet?.Datasets == null || !nextSet.Datasets.Any())
+                {
+                    throw new KeenException("Failed to fetch definition list");
+                }
+
+                allDefinitions.AddRange(nextSet.Datasets);
+
+            } while (firstSet.Count > allDefinitions.Count);
```

Comment:
This line seems confusing without a comment, but I plan to change how this works anyway.

Reply:
I just wonder what happens if the first request can give us a `Count` that can change by the time we request a later page...say I request 10 defs, the `Count` is 15, so we request another 10 or 5 or whatever, but since then another couple datasets have been added. We could potentially stop if we get 20 since it's >15, even if there are now actually 30 datasets because a bunch got added.

Likewise, if we request 10, the `Count` is 15, and before we request another page, some datasets get deleted, then we request another 10, but since 5 have been deleted, we get 0 back...according to the check above we'll throw even though this is technically correct.

I'll probably open a bug to handle these two cases since they'll require some deeper testing and knowing how the back end will actually behave, plus they are somewhat corner cases, so aren't on the critical path to getting this functionality usable.

Reply:
See Issue #132